Optionally show the text field length but not the max

sjmcdowall opened the pull request:
If maxLength is set to TextField.noMaxLength (pow(2,59)) then continue to show the character counter but do not show the max length.

sjmcdowall committed:
Optionally show the text field length but not the max

If maxLength is set to TextField.noMaxLength (pow(2,59)) then continue to show the character counter but do not show the max length.

jonahwilliams commented:
Please revert the unrelated formatting changes, it will make this impossible to review

sjmcdowall commented:
@jonahwilliams -- Yeah, was wondering about that -- I am using prettier in VS Code .. ah .. any idea of an easy way to do that? :)  Perhaps there should be a .prettierrc type file in the project to specify to NOT do anything :)

sjmcdowall committed:
Revert "Optionally show the text field length but not the max"

This reverts commit 759b1a454ed45a21791221761715e7d4358809e2.

sjmcdowall committed:
Add ability to control the showing of maxLength

Added code and test code to optionally allow someone to redact the showing of the "/${widget.maxLength}" from the display by setting the maxLength to the special value TextField.noMaxLength .

sjmcdowall commented:
ok @jonahwilliams and @HansMuller -- cleaned it up to remove the whitespace formatting... :)

sjmcdowall commented:
So what's the next steps here/??

jonahwilliams commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -191,6 +194,10 @@ class TextField extends StatefulWidget {
+  
+  /// If [maxLength] is set to this number, it disables the showing of the maximum length part of
+  /// the display, yet still gives a character counter.
+  static const int noMaxLength = 9007199254740992; // math.pow(2, 53);
```
This doesn't seem like the right approach

sjmcdowall replied:
@jonahwilliams -- It was the approach suggested by @HansMuller in the Issue:

[Issue](https://github.com/flutter/flutter/issues/23791)

jonahwilliams replied:
sounds good then

jonahwilliams commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -339,7 +346,10 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
 
     final int currentLength = _effectiveController.value.text.runes.length;
-    final String counterText = '$currentLength/${widget.maxLength}';
+    final String counterText = '$currentLength' +
```
This will also either require accessibility localizations

sjmcdowall replied:
Ah -- I have no idea what "accessibility localizations" actually me since the old code obviously didn't have them either. Do you have a pointer as to what that is?

jonahwilliams replied:
Currently when talkback is enabled we read out the the number of remaining characters. It doesn't make sense to do that if there is no limit, so you'll want to make sure semanticCounterText is empty

jonahwilliams commented:
cc @willlarche for material

jonahwilliams reviewed (commented):
Please also fix the analysis warnings

sjmcdowall commented:
@jonahwilliams -- How do I see the analyze failures you are referring to?  When I click on the analyze "Details" in the checks area -- it just hangs forever before return an error ...  ??

sjmcdowall committed:
Remove trailing whitespace

Self explanatory --

sjmcdowall committed:
And more trailing whitespace

Guess we had some in the test file too..

sjmcdowall committed:
Yikes -- whitespace is insidious

Yet more whitespace I missed... sigh..

sjmcdowall committed:
Fixes as per PR comments

Fixes:
-- Un-needed DefaultTextStyle in tests
-- Comment too long
-- semanticCounterText being set when noMaxLength used..

sjmcdowall commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -339,7 +346,10 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
+        ((widget.maxLength == TextField.noMaxLength)
+            ? ''
+            : '/${widget.maxLength}');
     final int remaining = (widget.maxLength - currentLength).clamp(0, widget.maxLength);
```
Fixed in latest Push

sjmcdowall commented on packages/flutter/test/material/text_field_test.dart:
```diff
@@ -1808,6 +1808,30 @@ void main() {
+    await tester.pumpWidget(const MaterialApp(
+      home: Material(
+        child: DefaultTextStyle(
+          style: TextStyle(fontFamily: 'Ahem', fontSize: 10.0),
```
Fixed in this test and I took the opportunity to "fix" it elsewhere .. tests still run correctly :)

sjmcdowall commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -192,6 +195,10 @@ class TextField extends StatefulWidget {
   /// {@macro flutter.widgets.editableText.maxLines}
   final int maxLines;
 
+  /// If [maxLength] is set to this number, it disables the showing of the maximum length part of
```
Assuming you mean **80** -- fixed :)

HansMuller replied:
Yes!

HansMuller commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -78,7 +78,10 @@ class TextField extends StatefulWidget {
   /// field, showing how many characters have been entered and how many are
-  /// allowed. After [maxLength] characters have been input, additional input
+  /// allowed unless the value is set to [noMaxLength] in which case only the
+  /// current length is displayed and the maximum size is redacted.
```
Just "only the current length is displayed" makes the point well enough.

HansMuller commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -192,6 +195,11 @@ class TextField extends StatefulWidget {
   /// {@macro flutter.widgets.editableText.maxLines}
   final int maxLines;
 
+  /// If [maxLength] is set to this number, it disables the showing
```
If [maxLength] is set to this value, only the "current input length" part of the character counter is shown.

HansMuller commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -339,9 +347,19 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
+    String counterText = '$currentLength';
+    String semanticCounterText = '';
+
+    // See if we are showing the maxLength or hiding it and therefore
```
This comment probably isn't needed, the code speaks for itself here.

HansMuller commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -339,9 +347,19 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
+    // See if we are showing the maxLength or hiding it and therefore
+    // if not hiding the semanticCounterText should be the remaining length
+    if (widget.maxLength != TextField.noMaxLength) {
+      // Show the maxLength on display and semanticCounterText
```
Similarly, this comment probably isn't needed

HansMuller reviewed (approved):
Thanks for persevering!  This version of the PR looks good. I've left a few suggestions about the docs which I hope you'll incorporate. Other than that it's ready to land.

LGTM

sjmcdowall committed:
Comment editing..

Clean up some comments and remove unneeded ones..

sjmcdowall commented:
No worries - it was fun!

Hixie commented:
Is there any reason we couldn't use `null` instead of an arbitrary number to represent "no limit"?

sjmcdowall commented:
@Hixie  -- null already has a meaning if you look at the code and doc.  It means no restriction ... and do not display anything ...

Hixie commented:
Ah, ok. How about -1? That at least couldn't possibly be a valid value, right?

sjmcdowall commented:
Well -- damn -- that's pretty damn clever @Hixie  -- I followed this approach based on a conversation with @HansMuller  -- but damn . Why the heck didn't I think of using -1 ??

HansMuller commented:
I thought `noMaxLength`, whose value is an approximation of `int.infinity`, was more evocative than -1. It also works as an actual length limit, so there wasn't a need to special case its value.

sjmcdowall commented:
So is this discussion completed @HansMuller and @Hixie ??

Hixie commented:
The problem with 9007199254740992 is that 9007199254740993 is a valid number in Flutter, but not in Hummingbird, which is really confusing. Something like -1 would be unambiguous everywhere. The other problem with 9007199254740992 is that if you end up printing it for some reason, it looks like magic or like a bug, whereas -1 would be clearly a sentinel value.

sjmcdowall commented:
I will happily change it to -1 if that is the official decision and you can get @HansMuller on board .. I sort of assume you are both on the Google Flutter team?!  I know I got a sign off from Will Larch @willlarche of Material -- but that maybe more on the general idea and UI aspect, not the UX.  So --- just let me know what to do :)

willlarche commented:
Seems like a valid use case even if we don’t make it an official one. :)
